### ISSUE TITLE
Test parallelization pending specs

### ISSUE THREAD

**Issue by jesusvazquez** [CartoDB/cartodb#5906]
I've been working in test parallelization for a while. Right now we can actually start a parallel execution running the specs in batches with **gnu parallel**, specifying the batch size with the -j (jobs/threads) parameter.

The batch size is limited by the machine processor in terms of performance. I'm using a 4 core droplet running the specs in 12 threads batch size and it takes **10-11 minutes to run all the specs**. That being said, running this with a better machine with more cores means less time.

First workflow was defining the batch size, for example 12 and then creating 12 databases and starting 12 redis but some specs failed randomly because they were using non empty databases. I had to find some deterministic results so I had to change the workflow.

Right now I'm generating the speclist by performing some cat,greps and seds to the Makefile and after that I count the lines. I create as much databases as specs we have to resolve the problem I had before. Then I define the batch size, for example 12 and I start 12 redis servers in different ports.

I've been testing this with multiple executions and the results are now more deterministic. By now we have this errors:

// DEPRECATED
- [X] spec/integrations/common_data_integration.rb [5907](https://github.com/CartoDB/cartodb/issues/5907) ( @gfiorav )
- [X] spec/models/visualization/collection_spec.rb ( @Kartones )
- [ ] spec/requests/carto/api/layers_controller_spec.rb [5908](https://github.com/CartoDB/cartodb/issues/5908) ( @Kartones )
- [X] spec/lib/trending_maps_spec.rb 

Random errors
- Errno::EADDRINUSE: [5909](https://github.com/CartoDB/cartodb/issues/5909) (@gfiorav)
  - [x] spec/requests/api/imports_spec.rb
  - [x] spec/requests/carto/api/imports_controller_spec.rb 
  - [x] spec/models/asset_spec.rb
- Others 
  - [ ] services/importer/spec/unit/runner_spec.rb 

//

cc/ @luisbosque @Kartones @juanignaciosl @ethervoid

**Comment by Kartones** [comment]
Assign to me at the second and third ones (vis collection and layers controller), but I'll need the branch and how to run them to reprouce the error, but for sure I'll stabilize those.

Data imports one will need some changes in how it fakely uploads files, but if @rafatower team is too busy assign to me too, as this is important to start saving precious time running tests

**Comment by jesusvazquez** [comment]
The branch is named ```test-improvements``` and it actually has a few scripts in the root to make it easier but they will be placed somewhere else in the future.

In order to run this locally you have to install gnu parallel in your machine. I don't know if its actually in the ubuntu repositories but you can do it easily with the following steps:
```
wget http://ftp.gnu.org/gnu/parallel/parallel-20130622.tar.bz2
tar -xvf parallel-20130622.tar.bz2 
cd parallel-20130622/
sudo ./configure && sudo make && sudo make install
```

And test it with this command ```parallel echo ::: A B C````that should return 
```
ubuntu@test01:~$ parallel echo ::: A B C
A
B
C
```

After this you only need to checkout the branch. To run all of them in parallel you need to generate the specfull.txt with the following command
```
ubuntu@test07:~$ sh generateSpecFull.sh 
# Speclist has been created
```
Then you need to start the redis and create the databases with the following command.
```
ubuntu@test07:~/www/production.cartodb.com/rsync$ sh wrapper.sh 20
# Wrapper finished
```
**Note**: the number 20 is the number of threads that we're going to use to execute all the tests. For safety reasons we're going to do it with 18 threads. 

To start running all the tests execute this:
```
date; time parallel -j 18 --delay 0.1 -a specfull.txt './executor.sh'; date
```

-----
Then if you want to run just one spec specifying the database and the redis you can run this:
```

**Comment by ethervoid** [comment]
We should start using this http://apidock.com/rails/ActionController/TestProcess/fixture_file_upload to fake the file uploads using fixtures

**Comment by jesusvazquez** [comment]
cc/ adding @gfiorav because he wrote spec/integrations/common_data_integration.rb we will check this error on tuesday.

**Comment by jesusvazquez** [comment]
Added two new conflictive specs.
  - trending_map_spec seems to be a new error related with visualizations.
  - runner.spec
I'll work with this two tomorrow.

**Comment by jesusvazquez** [comment]
retest this please

**Comment by juanignaciosl** [comment]
Please let me know if I can help with this.

**Comment by rafatower** [comment]
is this ticket complete?

**Comment by gfiorav** [comment]
@jesusvazquez was testing it thoroughly. I think no problems so far.

**Comment by jesusvazquez** [comment]
The asset spec seems to be OK but sometimes it remains IDLE and never ends. This is a random issue with low priority. 

If this happens you can either kill the job or kill the process inside the testing server and it will finish. Both are manual and not ideal "solutions". 

I created a pull request in order to deactivate the spec in the parallel execution and run it later in serial with the other failed specs. Again this is a temporal solution until we dig in and fix the issue.

The tests that have the same condition are the following:

- [ ] spec/integrations/common_data_integration.rb
- [ ] spec/models/visualization/table_blender.rb
- [ ] spec/models/asset_spec.rb
- [ ] services/user-mover/spec/user_mover_spec.rb
- [ ] spec/requests/carto/api/layers_controller_spec.rb

**Comment by rafatower** [comment]
do you have any more clue about why they fail? a lock in db maybe?

the user mover ones strike as important to me.

**Comment by jesusvazquez** [comment]
It does request futher research in the logs for each execution. I'll keep this issue updated and will post more details for each spec. Maybe the user spec is a good start but I want to clarify again that the user spec is running okey sequentially.

**Comment by alonsogarciapablo** [comment]
There has been no activity on this issue for more several months. We are closing it. If you think this still needs to be addressed please open a new issue.
